ecosystem/indexer/src/indexer/tailer.rs

```diff
@@ -0,0 +1,416 @@
+            tasks.push(task);
+        }
+        let results = await_tasks(tasks).await;
+        results
+    }
+
+    pub async fn process_transaction(
+        &self,
+        txn: Arc<Transaction>,
+    ) -> anyhow::Result<Vec<Result<ProcessingResult, TransactionProcessingError>>> {
+        let mut tasks = vec![];
+        for processor in &self.processors {
+            let txn2 = txn.clone();
+            let processor2 = processor.clone();
+            let task = tokio::task::spawn(async move {
+                processor2.process_transaction_with_status(txn2).await
+            });
+            tasks.push(task);
+        }
+        let results = await_tasks(tasks).await;
+        Ok(results)
+    }
+
+    pub async fn get_next_txn(&mut self) -> Arc<Transaction> {
+        Arc::new(self.transaction_fetcher.lock().await.fetch_next().await)
+    }
+
+    pub async fn get_txn(&self, version: u64) -> Arc<Transaction> {
+        Arc::new(
+            self.transaction_fetcher
+                .lock()
+                .await
+                .fetch_version(version)
+                .await,
+        )
+    }
+}
+
+pub async fn await_tasks<T: Debug>(tasks: Vec<JoinHandle<T>>) -> Vec<T> {
+    let mut results = vec![];
+    for task in tasks {
+        let result = task.await;
+        if result.is_err() {
+            aptos_logger::error!("Error joining task: {:?}", &result);
+        }
+        results.push(result.unwrap());
+    }
+    results
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::{
+        database::{new_db_pool, PgPoolConnection},
+        default_processor::DefaultTransactionProcessor,
+        models::transactions::TransactionModel,
+    };
+    use diesel::Connection;
+    use serde_json::json;
+
+    pub fn wipe_database(conn: &PgPoolConnection) {
+        for table in [
+            "events",
+            "user_transactions",
+            "block_metadata_transactions",
+            "transactions",
+            "processor_statuses",
+            "__diesel_schema_migrations",
+        ] {
+            conn.execute(&format!("DROP TABLE IF EXISTS {}", table))
+                .unwrap();
+        }
+    }
+
+    pub fn setup_indexer() -> anyhow::Result<(PgDbPool, Tailer)> {
+        let database_url = std::env::var("INDEXER_DATABASE_URL")
+            .expect("must set 'INDEXER_DATABASE_URL' to run tests!");
+        let conn_pool = new_db_pool(database_url.as_str())?;
+        wipe_database(&conn_pool.get()?);
+
+        let mut tailer = Tailer::new("http://fake-url.aptos.dev", conn_pool.clone())?;
+        tailer.run_migrations();
+
+        let pg_transaction_processor = DefaultTransactionProcessor::new(conn_pool.clone());
+        tailer.add_processor(Arc::new(pg_transaction_processor));
+        Ok((conn_pool, tailer))
+    }
+
+    #[tokio::test]
+    async fn test_parsing_and_writing() {
+        let (conn_pool, tailer) = setup_indexer().unwrap();
+        // An abridged genesis transaction
```

Comment:
I'm not comfortable adding in a test that will fail if you don't go through the hassle of doing a non-trivial setup. We have at least two such tests in the code base and they are all optional:
* prover
* vault

while some folks may only run tests in their own space for dev, this will be a huge pain in the ass for anyone doing multicrate development who turns no e2e testing, walks away, and then realizes that they forgot to exclude this one test.